File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void fun(string temp,string digits,string mapping[],vector<string> &ans){
        if(digits.size() == 0){
            ans.push_back(temp);
            return;
        }
        
        int ind = (int)(digits[0] - '0') - 2;
        for(int i = 0;i<mapping[ind].size();i++){
            char ch = mapping[ind][i];
            fun(temp+ch,digits.substr(1,digits.size()-1),mapping,ans);
            // list.push_back(arr);
            // fun(temp+ch,digits.substr(1,digits.size()-1),mapping);
        }
        // return list;
    }
    vector<string> letterCombinations(string digits) {
        
        vector<string> ans;
        if(digits.size() == 0){
            return ans;
        }
        string mapping[] = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        string temp = "";
        fun(temp,digits,mapping,ans);
        return ans;
        
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0){
            ArrayList<String> list = new ArrayList<String>();
            return list;
        }
        String[] mapping = new String[]{"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        // String temp = "";
        return fun("",digits,mapping);
    }
    public List<String> fun(String temp,String digits,String[] mapping){
        if(digits.length() == 0){
            ArrayList<String> list = new ArrayList<>();
            list.add(temp);
            return list;
        }
        ArrayList<String> list = new ArrayList<>();
        
        int ind = (int)(digits.charAt(0) - '0') - 2;
        
        for(int i = 0;i<mapping[ind].length();i++){
            char ch = (char)(mapping[ind].charAt(i));
            // System.out.println(temp+ch);
            list.addAll(fun(temp+ch,digits.substring(1),mapping));
        }
        return list;
    }
}